### PR TITLE
feat: add ci

### PR DESCRIPTION
**Pull request by adschwartz**
Add a readme and CI steps to publish to dockerhub.

### CODE REVIEW

**Comment by gbouv** [comment]
@adschwartz Let's merge the few circle optimization I did in this separate PR before we merge yours if you don't mind: #32

File: .circleci/config.yml
Comment (adschwartz) [inline .circleci/config.yml:171]: Ah I just realized that I actually only want to run this when the PR is merged. Perhaps either of you have an idea of how to configure the script for that? I'll do some more digging.

Comment (gbouv) [reply .circleci/config.yml:171]: @adschwartz what behaviour do you want exactly? I'm guessing maybe you want this job to run only on main (since you're publishing to docker), but at the same time you mention `branched > ignore > main`, so not sure.
If what you want is really a one-off thing to start running it when this PR merges, what I'd do would be to set a branch filter to run it ONLY on main for this PR, and then when this PR is merged, you can remove the filter to run it everywhere (or start ignoring main if you want to run it only on PR branches). At least that's how I'd do it

Comment (gbouv) [reply .circleci/config.yml:171]: That being said, I think a pretty common pattern is to split the "build" and "publish" phase of a job. Here your `build_deploy_simple_api` is doing both. I think if you split it, at least you'd be able to run the 'build' phase on PRs everywhere, and then you can activate the publish part either later, or only on main (i.e. which means it will start running when this PR is merged, but only on main).

File: .circleci/config.yml
Comment (gbouv) [inline .circleci/config.yml:140]: Interesting, we don't care about versioning here? Just to make sure we're on the same page, this will basically override the docker image everytime, trashing the previous version

File: .circleci/config.yml
Comment (gbouv) [inline .circleci/config.yml:177]: Hmm, so you're going to publish a new docker image on every single commit, is that what we want?
For example if we submit 2 different PRs touching this sub folder, the docker images on docker-hub will oscillate between each PR on every commit. This could be confusing, esp. since we're tagging all docker images with `latest`. Don't we want instead to publish a docker image on main only?

File: .circleci/config.yml
Comment (gbouv) [inline .circleci/config.yml:143]: Why do we have this line? I believe this isn't doing anything. You're basically running something like:
```
docker tag kurtosistech/kurtosis-simple-api:latest kurtosistech/kurtosis-simple-api:latest
```
If the image is already named `kurtosistech/kurtosis-simple-api:latest` I guess there's no need to tag it with that same name :)

Comment (gbouv) [reply .circleci/config.yml:143]: (in the monorepo we're doing `docker tag "${image_name_with_version}" "${image_name_no_version}"`) because we want the image we're publishing to have 2 tags, one `:latest` and one with the actual version, like `:0.67.2` whatever)

File: simple-api/README.md
Comment (gbouv) [inline simple-api/README.md:24]: Just noting here that we don't need `kurtosis.yml`, this should do the same:
```
kurtosis run kurtosis-package/ --enclave-id kurtosis-enclave
```

File: simple-api/README.md
Comment (gbouv) [inline simple-api/README.md:162]: Wait, is that true? Just to make sure I understand. Let's say I am in this repo and I run:
```
./script/build.sh
```
After that, I'll have a new image in my local docker named `kurtosistech/kurtosis-simple-api:latest`
If after that I run my starlark package, since the image in my docker already exist, the CLI will not bother downloading it from docker hub, it will just use the existing one.

I suppose you've tested that, so I'm wondering if there's a bug somewhere here. If I understand correctly, the behaviour you're describing here is really not what we want, so maybe there's a bug that needs to be fixed

File: simple-api/kurtosis-package/main.star
Comment (gbouv) [inline simple-api/kurtosis-package/main.star:8]: Haha, I added this line previously (and I think it was actually Gyani who asked me to add it). I think it's part of Python general style guide to separate root functions with 2 blank lines

**Review by gbouv — APPROVED** [review]
Approving to unblock, but there might be 2 or 3 things we want to quickly catch up on:
- what's the intend of the CI job w.r.t. publishing to docker hub? Here we're publishing on every commit, except on main, which I think is a bit of an anti-pattern. Just want to make sure you really want that
- On the readme, what you mention about Kurtosis CLI replacing a local version of the image with the one from docker hub I believe is incorrect, unless I misunderstood. Might be worth clarifying to avoid confusion for the candidates

File: .circleci/config.yml
Comment (h4ck3rk3y) [reply .circleci/config.yml:171]: 💯  to what Guillaume said; could use https://github.com/kurtosis-tech/kurtosis/blob/main/.circleci/config.yml for inspiration (though in this case publish happens on tag; you can change it to publish happening on merge to main)

build - the ignore main runs on PRs and random branches
publish - while the one set to run on main only run when the merge happens (as thats the only way to write to main)

File: .circleci/config.yml
Comment (h4ck3rk3y) [reply .circleci/config.yml:140]: and if this happens with every push to any PR then its quite risky as someone might publish something that doesn't work and then the user (interview candidate in this case is left with a broken thing)

File: README.md
Comment (h4ck3rk3y) [inline README.md:7]: golang -> Golang

File: simple-api/README.md
Comment (h4ck3rk3y) [inline simple-api/README.md:64]: as this -> like this

File: simple-api/README.md
Comment (h4ck3rk3y) [inline simple-api/README.md:94]: correction here; its running on 8080 on the host on which the container is running

within the enclave itself you can have multiple things that look like http: 8080/tcp ( the mapping being on the container and nothing to do with the enclave itself)

File: simple-api/README.md
Comment (h4ck3rk3y) [inline simple-api/README.md:138]: They also need to know the containers internal ip address for 8080 to work

**Review by h4ck3rk3y — COMMENTED** [review]
just went through this quickly as @gbouv had already reviewed it!

File: simple-api/README.md
Comment (adschwartz) [reply simple-api/README.md:162]: haha @gbouv I did test it and this was the behavior I observed. I assumed that this was the way it was supposed to work, but also struggled with coming to terms with that since it seems counterproductive. Maybe we should look at this together to make sure I'm indeed doing this right?

File: simple-api/kurtosis-package/main.star
Comment (adschwartz) [reply simple-api/kurtosis-package/main.star:8]: Oh cool. I'll add it back haha.

File: simple-api/README.md
Comment (adschwartz) [reply simple-api/README.md:94]: Good point I'll add that

File: .circleci/config.yml
Comment (adschwartz) [reply .circleci/config.yml:171]: @gbouv My intention was to only build and publish on merge to main. But I can see the value in building if there's a PR and then build and publish on merge to main. I was looking for examples of us doing this but didn't find any. As @h4ck3rk3y mentioned we run jobs based on the presence of tags I believe. I was googling this a bit yesterday and didn't find a simple way to filter on if there's a merge or not. But I could be overlooking it and was looking to see if you two knew how to do it.

Comment (h4ck3rk3y) [reply .circleci/config.yml:171]: Any push to main is a merge as our repository is protected. I'd replace the tags in the publish jobs to main like 

```
      - publish_docker_image
          context:
            - docker-user
          filters:
            branches:
              only: "main"
```

File: .circleci/config.yml
Comment (adschwartz) [reply .circleci/config.yml:140]: I had actually chatted with @h4ck3rk3y about this a while back and at the time we felt it was overkill to have versioning for this image based on what we are doing. Personally I still think it is, but I'm willing to be convinced otherwise if you both feel strongly. This image is an example that we will likely use in our FE candidate interviewing.

If you break the image the user will be stuck with a broken image. That seems like a general truth though (i.e. applies everywhere). If we use this with candidates I doubt they would just use an earlier version. We should make sure that the images we are pushing work. But that's just my $0.02. Willing to have my mind changed if you feel differently :)

Comment (h4ck3rk3y) [reply .circleci/config.yml:140]: We don't explicitly need versioning but we shouldn't be publishing every commit that gets made to any PR on this project as that commit might be broken & the image would get in a bad state.

I'd expect a merge to main to get published and you can tag it as latest & if you want you can also add the latest git commit tag using (git rev-parse --short HEAD)

File: .circleci/config.yml
Comment (gbouv) [reply .circleci/config.yml:171]: @adschwartz let me know if I'm missing something but I think you can achieve this with something like:
```
jobs:
  build_simple_api:
    executor: ubuntu_vm
    steps:
      - checkout
      # If this sub-repo hasn't changed, don't run the tests
      - run: |
          if git --no-pager diff --exit-code origin/main...HEAD -- ':.circleci' ':simple-api'; then
            circleci-agent step halt
          fi
      - run: simple-api/scripts/build.sh

  publish_simple_api:
    executor: ubuntu_vm
    steps:
      - checkout
      # If this sub-repo hasn't changed, don't run the tests
      - run: |
          if git --no-pager diff --exit-code origin/main...HEAD -- ':.circleci' ':simple-api'; then
            circleci-agent step halt
          fi
      - run: simple-api/scripts/build.sh
      - run: echo "${DOCKER_PASSWORD}" | docker login -u ${DOCKER_USERNAME} --password-stdin
      - run: |
          set -euo pipefail
          version_to_publish="latest"
          image_org_and_repo="kurtosistech/kurtosis-simple-api"
          image_name_with_version="${image_org_and_repo}:${version_to_publish}"
          docker push "${image_name_with_version}"
workflows:
  build:
    jobs:
      - build_simple_api:
          filters:
            branches:
              ignore:
                - main
      - publish_simple_api:
          filters:
            branches:
              only:
                - main
```

File: .circleci/config.yml
Comment (gbouv) [reply .circleci/config.yml:140]: Yes, agreed with Gyani, if we're only publishing on the main branch I'm fine keeping the tag to `latest` for every publish

File: simple-api/README.md
Comment (gbouv) [reply simple-api/README.md:162]: Ohh, then I think you're onto something @adschwartz, this definitely sounds like a bug. I'm happy to quickly look into it if you don't want to allocate some of your time to this specific issue. BUt also happy to look into it together if you're curious. If the former feel free to create a ticket an assign it to me, otherwise just ping me and we can quickly chat about it

Comment (gbouv) [reply simple-api/README.md:162]: But I think anyway you should rephrase that to be in sync with the desired behaviour (i.e. we pull local images first, and fallback to docker hub download if not found), such that you don't have to update the wording here once the bug is fixed

File: .circleci/config.yml
Comment (adschwartz) [reply .circleci/config.yml:171]: @h4ck3rk3y @gbouv wow you guys rock. Didn't expect you to hand me the solution, but really appreciate it!

File: .circleci/config.yml
Comment (adschwartz) [reply .circleci/config.yml:140]: Oh I might add the git commit tag as well. Like that idea. thank you both!

File: .circleci/config.yml
Comment (adschwartz) [reply .circleci/config.yml]: Thank you both for the inputs here. Appreciated. @gbouv @h4ck3rk3y

**Pull request merged**
